core/src/Kokkos_HostSpace.hpp

```diff
@@ -183,29 +183,24 @@ namespace Kokkos {
 
 namespace Impl {
 
-template <>
-struct DeepCopy<HostSpace, HostSpace, DefaultHostExecutionSpace> {
-  DeepCopy(void* dst, const void* src, size_t n) {
-    hostspace_parallel_deepcopy(dst, src, n);
-  }
-
-  DeepCopy(const DefaultHostExecutionSpace& exec, void* dst, const void* src,
-           size_t n) {
-    hostspace_parallel_deepcopy_async(exec, dst, src, n);
-  }
-};
-
 template <class ExecutionSpace>
 struct DeepCopy<HostSpace, HostSpace, ExecutionSpace> {
   DeepCopy(void* dst, const void* src, size_t n) {
     hostspace_parallel_deepcopy(dst, src, n);
   }
 
   DeepCopy(const ExecutionSpace& exec, void* dst, const void* src, size_t n) {
-    exec.fence(
-        "Kokkos::Impl::DeepCopy<HostSpace, HostSpace, "
-        "ExecutionSpace>::DeepCopy: fence before copy");
-    hostspace_parallel_deepcopy_async(dst, src, n);
+    // If ExecutionSpace is a Device ExecutionSpace, we should call
+    // this from DefaultHostExecutionSpace
+    if constexpr (std::is_same_v<ExecutionSpace, DefaultExecutionSpace> &&
+                  !std::is_same_v<ExecutionSpace, DefaultHostExecutionSpace>) {
```

Comment:
I find it hard to argue about this. Do you mind using `SpaceAccessibility` instead?
```suggestion
    if constexpr (!Kokkos::SpaceAccesibility<ExecutionSpace, Kokkos::HostSpace>::accessible) {
```

Reply:
This is indeed a better way. I have fixed accordingly.